File: spring/src/main/java/cardsproject/domain/content/DraftSession.java
package cardsproject.domain.content;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "draft_sessions")
public class DraftSession {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    private DraftSessionStatusType status;
    @Enumerated(EnumType.STRING)
    private DraftSessionDraftTypeType draftType;
    private Integer seats = 8;
    private Integer timePerPickSeconds = 30;
    private LocalDateTime createdAt;
    private LocalDateTime completedAt;

    @Column(name = "card_set_id")
    private Long cardSetId;

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public DraftSessionStatusType getStatus() { return status; }
    public void setStatus(DraftSessionStatusType status) { this.status = status; }
    public DraftSessionDraftTypeType getDraftType() { return draftType; }
    public void setDraftType(DraftSessionDraftTypeType draftType) { this.draftType = draftType; }
    public Integer getSeats() { return seats; }
    public void setSeats(Integer seats) { this.seats = seats; }
    public Integer getTimePerPickSeconds() { return timePerPickSeconds; }
    public void setTimePerPickSeconds(Integer timePerPickSeconds) { this.timePerPickSeconds = timePerPickSeconds; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public LocalDateTime getCompletedAt() { return completedAt; }
    public void setCompletedAt(LocalDateTime completedAt) { this.completedAt = completedAt; }
    public Long getCardSetId() { return cardSetId; }
    public void setCardSetId(Long cardSetId) { this.cardSetId = cardSetId; }

    // ── Business operations ──────────────────────────────────────────
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void start() {
        // TODO: implement start
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void abandon() {
        // TODO: implement abandon
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void complete() {
        // TODO: implement complete
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public Boolean isFull() {
        // TODO: implement isFull
        return null;
    }

    // ── Validation rules ─────────────────────────────────────────────
    @jakarta.validation.constraints.AssertTrue(message = "Draft session must have between 2 and 16 seats")
    @com.fasterxml.jackson.annotation.JsonIgnore
    public boolean isSeatsRangeValid() {
        return (getSeats() == null || (getSeats() >= 2 && getSeats() <= 16));
    }
    @jakarta.validation.constraints.AssertTrue(message = "Time per pick must be greater than zero")
    @com.fasterxml.jackson.annotation.JsonIgnore
    public boolean isTimePerPickPositiveValid() {
        return (getTimePerPickSeconds() == null || getTimePerPickSeconds() > 0);
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static final java.util.Map<DraftSessionStatusType, java.util.List<DraftSessionStatusType>> ALLOWED_TRANSITIONS =
        java.util.Map.ofEntries(
        java.util.Map.entry(DraftSessionStatusType.WAITINGFORPLAYERS, java.util.List.of(DraftSessionStatusType.DRAFTING, DraftSessionStatusType.ABANDONED)),
        java.util.Map.entry(DraftSessionStatusType.DRAFTING, java.util.List.of(DraftSessionStatusType.COMPLETED, DraftSessionStatusType.ABANDONED))
        );

    public void assertTransition(DraftSessionStatusType to) {
        java.util.List<DraftSessionStatusType> allowed = ALLOWED_TRANSITIONS.getOrDefault(this.getStatus(), java.util.List.of());
        if (!allowed.contains(to)) {
            throw new IllegalStateException("Transition " + this.getStatus() + " -> " + to + " not allowed");
        }
    }
}

File: symfony/src/Entity/Content/DraftSession.php
<?php

namespace App\Entity\Content;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;
use Symfony\Component\Serializer\Annotation\SerializedName;
use App\Repository\Content\DraftSessionRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use App\Entity\Cards\CardSet;

#[ORM\Entity(repositoryClass: DraftSessionRepository::class)]
#[ORM\Table(name: 'draft_session')]
class DraftSession
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    #[Groups(['draftSession:read'])]
    private ?int $id = null;

    #[ORM\Column(type: 'string', length: 20)]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private string $status = 'WaitingForPlayers';

    #[ORM\Column(type: 'string', length: 20)]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private string $draftType = 'Booster';

    #[ORM\Column(type: 'json', nullable: true)]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private ?array $packContents = null;

    #[ORM\Column(type: 'integer')]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private int $seats = 8;

    #[ORM\Column(type: 'integer')]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private int $timePerPickSeconds = 30;

    #[ORM\Column(type: 'datetime', nullable: true)]
    #[SerializedName('createdAt')]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private ?\DateTimeInterface $createdAt = null;

    #[ORM\Column(type: 'datetime', nullable: true)]
    #[SerializedName('completedAt')]
    #[Groups(['draftSession:read', 'draftSession:write'])]
    private ?\DateTimeInterface $completedAt = null;

    #[ORM\ManyToOne(targetEntity: CardSet::class, inversedBy: 'draftSessions')]
    #[ORM\JoinColumn(nullable: false, onDelete: 'RESTRICT')]
    private ?CardSet $cardSet = null;

    #[ORM\OneToMany(mappedBy: 'session', targetEntity: DraftParticipant::class)]
    private Collection $participants;

    public function __construct()
    {
        $this->participants = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getStatus(): string
    {
        return $this->status;
    }

    public function setStatus(string $status): static
    {
        $this->status = $status;
        return $this;
    }

    public function getDraftType(): string
    {
        return $this->draftType;
    }

    public function setDraftType(string $draftType): static
    {
        $this->draftType = $draftType;
        return $this;
    }

    public function getPackContents(): ?array
    {
        return $this->packContents;
    }

    public function setPackContents(?array $packContents): static
    {
        $this->packContents = $packContents;
        return $this;
    }

    public function getSeats(): int
    {
        return $this->seats;
    }

    public function setSeats(int $seats): static
    {
        $this->seats = $seats;
        return $this;
    }

    public function getTimePerPickSeconds(): int
    {
        return $this->timePerPickSeconds;
    }

    public function setTimePerPickSeconds(int $timePerPickSeconds): static
    {
        $this->timePerPickSeconds = $timePerPickSeconds;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    public function setCreatedAt(?\DateTimeInterface $createdAt): static
    {
        $this->createdAt = $createdAt;
        return $this;
    }

    public function getCompletedAt(): ?\DateTimeInterface
    {
        return $this->completedAt;
    }

    public function setCompletedAt(?\DateTimeInterface $completedAt): static
    {
        $this->completedAt = $completedAt;
        return $this;
    }

    #[Groups(['draftSession:read'])]
    public function getCardSetId(): ?int
    {
        return $this->cardSet?->getId();
    }

    public function getCardSet(): ?CardSet
    {
        return $this->cardSet;
    }

    public function setCardSet(?CardSet $cardSet): static
    {
        $this->cardSet = $cardSet;
        return $this;
    }

    public function getParticipants(): Collection
    {
        return $this->participants;
    }

    // ── Validation rules ─────────────────────────────────────────────
    #[\Symfony\Component\Validator\Constraints\IsTrue(message: "Draft session must have between 2 and 16 seats")]
    public function isSeatsRangeValid(): bool
    {
        return ($this->getSeats() === null || ($this->getSeats() >= 2 && $this->getSeats() <= 16));
    }

    #[\Symfony\Component\Validator\Constraints\IsTrue(message: "Time per pick must be greater than zero")]
    public function isTimePerPickPositiveValid(): bool
    {
        return ($this->getTimePerPickSeconds() === null || $this->getTimePerPickSeconds() > 0);
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->getCompletedAt() !== null && !($this->getStatus() === 'COMPLETED')) {
            throw new \DomainException('completed_at can only be set when draft status is Completed');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function start(): void
    {
        // TODO: implement start
    }

    public function abandon(): void
    {
        // TODO: implement abandon
    }

    public function complete(): void
    {
        // TODO: implement complete
    }

    public function isFull(): mixed
    {
        // TODO: implement is_full
        return null;
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static array $ALLOWED_TRANSITIONS = [
        'WaitingForPlayers' => ['Drafting', 'Abandoned'],
        'Drafting' => ['Completed', 'Abandoned'],
    ];

    public function assertTransition(string $from, string $to): void
    {
        $allowed = self::$ALLOWED_TRANSITIONS[$from] ?? [];
        if (!in_array($to, $allowed, true)) {
            throw new \RuntimeException("Transition {$from} -> {$to} not allowed");
        }
    }

}
